File: scripts/v5/o5f-target-tier.mjs
#!/usr/bin/env node
/**
 * O5F §十B -- read the LIVE Target runtime's spectral sample tier.
 *
 * The question is what the Target actually compiled in the EXACT context our
 * 390x844 captures ran -- not what a viewport rule would predict. §十B is
 * explicit: do not infer from viewport dimensions. So each probe context is
 * built with the same settings o5r-measure.mjs's targetLane used (mobile
 * emulation exactly when w or h < 500), the same media routes, and the tier
 * is read from three independent places:
 *
 *   1. the compiled programs themselves -- GPUDevice.createShaderModule is
 *      hooked before any page script runs, and the card body program is the
 *      module that calls refract(); its refract count IS the sample count
 *      (the O5 compiled audit established exactly this equivalence on our
 *      own programs: five at high, three at low);
 *   2. the device-tier predicate's inputs, evaluated live in the page:
 *      matchMedia("(pointer: coarse)"), hardwareConcurrency, deviceMemory;
 *   3. the bundle the live page actually loaded -- every response carrying
 *      "maxDispersionSamples" is hashed so the scorer can prove the live
 *      runtime is the archived bundle the source contract was read from.
 *
 * Output: artifacts/optics-o5f/target-tier/target-tier-raw.json
 * (scored by o5f-target-tier.py into qa-v5/optics-o5f/target-mobile-tier.json)
 *
 * Usage: o5f-target-tier.mjs [--target=<url>] [--out=<dir>]
 */
import { createHash } from "node:crypto";
import { mkdir, writeFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { chromium } from "playwright";
import {
  installTargetRoutes, loadAsset, TARGET_VIDEO_HOOK, freezeTarget,
} from "./o2_media_routes.mjs";

const REPO = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");
const opts = {
  target: "https://infinite-liquid-glass.shader.se/?v=2",
  asset: "bw-split",
  out: path.join(REPO, "artifacts/optics-o5f/target-tier"),
};
for (const a of process.argv.slice(2)) {
  const [k, v] = a.replace(/^--/, "").split("=");
  if (k in opts) opts[k] = v;
}

const VIEWPORTS = [[1440, 900], [390, 844], [844, 390], [700, 700]];
const sha256 = (b) => createHash("sha256").update(b).digest("hex");

// Hooked before any page script. WGSL first (the Target runs three's WebGPU
// renderer, like us); the WebGL hook is the fallback so a driver that fell
// back to WebGL still yields programs rather than an empty probe.
const SHADER_HOOK = `(() => {
  window.__o5fShaders = [];
  window.__o5fEnvAtLoad = {
    pointerCoarse: window.matchMedia("(pointer: coarse)").matches,
    anyPointerCoarse: window.matchMedia("(any-pointer: coarse)").matches,
    hardwareConcurrency: navigator.hardwareConcurrency ?? null,
    deviceMemory: navigator.deviceMemory ?? null,
    maxTouchPoints: navigator.maxTouchPoints ?? null,
    userAgent: navigator.userAgent,
    webgpu: !!navigator.gpu,
  };
  if (window.GPUDevice) {
    const orig = GPUDevice.prototype.createShaderModule;
    GPUDevice.prototype.createShaderModule = function (desc) {
      try { window.__o5fShaders.push(String(desc && desc.code || "")); }
      catch (e) { /* never break the page */ }
      return orig.call(this, desc);
    };
  }
  for (const C of [window.WebGLRenderingContext, window.WebGL2RenderingContext]) {
    if (!C) continue;
    const orig = C.prototype.shaderSource;
    C.prototype.shaderSource = function (shader, source) {
      try { window.__o5fShaders.push(String(source)); }
      catch (e) { /* never break the page */ }
      return orig.call(this, shader, source);
    };
  }
})();`;

const browser = await chromium.launch({ channel: "chrome", headless: true,
  args: ["--enable-unsafe-webgpu", "--enable-webgpu-developer-features"] });
await mkdir(opts.out, { recursive: true });
const records = [];

for (const [w, h] of VIEWPORTS) {
  const tag = `${w}x${h}`;
  const mobile = w < 500 || h < 500;   // o5r-measure's own rule, unchanged
  const ctx = await browser.newContext({ viewport: { width: w, height: h },
    hasTouch: mobile, ...(mobile ? { isMobile: true } : {}) });
  await installTargetRoutes(ctx, loadAsset(opts.asset), null);
  await ctx.addInitScript(TARGET_VIDEO_HOOK);
  await ctx.addInitScript(SHADER_HOOK);
  const page = await ctx.newPage();
  const errors = [];
  page.on("pageerror", (e) => errors.push(String(e)));
  const bundles = [];
  page.on("response", async (resp) => {
    try {
      const url = resp.url();
      if (!/\.js(\?|$)/.test(url)) return;
      const body = await resp.body();
      if (body.includes("maxDispersionSamples")) {
        bundles.push({ url, bytes: body.length, sha256: sha256(body) });
      }
    } catch { /* detached responses at teardown are fine */ }
  });

  let loaded = true;
  try {
    await page.goto(opts.target, { waitUntil: "load", timeout: 60000 });
    await page.waitForFunction(() =>
      (window.__o2vids?.length ?? 0) >= 8
      && window.__o2vids.every((v) => v.readyState >= 2), undefined,
      { timeout: 90000 });
    await page.waitForTimeout(4000);
    await freezeTarget(page, 4);
  } catch (e) {
    loaded = false;
    errors.push(`load: ${String(e).slice(0, 200)}`);
  }

  const probe = loaded ? await page.evaluate(() => {
    const shaders = window.__o5fShaders ?? [];
    const refractModules = shaders
      .filter((s) => s.includes("refract"))
      .map((s) => ({
        length: s.length,
        refractCalls: (s.match(/refract/g) || []).length,
        textureSamples: (s.match(/textureSample\s*\(/g) || []).length
          + (s.match(/texture2D\s*\(|texture\s*\(/g) || []).length,
        isFragment: s.includes("@fragment") || s.includes("gl_FragColor")
          || s.includes("fragmentMain") || s.includes("main_fragment"),
        head: s.slice(0, 120),
      }));
    return {
      envAtLoad: window.__o5fEnvAtLoad ?? null,
      envNow: {
        pointerCoarse: window.matchMedia("(pointer: coarse)").matches,
        hardwareConcurrency: navigator.hardwareConcurrency ?? null,
        deviceMemory: navigator.deviceMemory ?? null,
      },
      shaderModuleCount: shaders.length,
      refractModules,
      videoCount: (window.__o2vids ?? []).length,
    };
  }) : null;

  // The card body program's full text, for the output-transform check (§十D)
  // and the archive: the module with the most refract() calls.
  let bodyProgram = null;
  if (loaded) {
    bodyProgram = await page.evaluate(() => {
      const shaders = (window.__o5fShaders ?? []).filter(
        (s) => s.includes("refract"));
      if (!shaders.length) return null;
      let best = shaders[0];
      for (const s of shaders) {
        if ((s.match(/refract/g) || []).length
            > (best.match(/refract/g) || []).length) best = s;
      }
      return best;
    });
    if (bodyProgram) {
      const f = `target-body-program-${tag}.wgsl.txt`;
      await writeFile(path.join(opts.out, f), bodyProgram);
      bodyProgram = { file: f, sha256: sha256(bodyProgram),
        length: bodyProgram.length };
    }
    const shot = `target-tier-${tag}.png`;
    await page.screenshot({ path: path.join(opts.out, shot) });
  }

  records.push({ vp: tag, mobileEmulation: mobile, loaded, probe,
    bodyProgram, bundlesWithPredicate: bundles,
    errorCount: errors.length, errors: errors.slice(0, 4) });
  console.log(`${tag}: loaded=${loaded} pointerCoarse=`
    + `${probe?.envAtLoad?.pointerCoarse} modules=${probe?.shaderModuleCount} `
    + `refractModules=${probe?.refractModules?.length}`);
  await ctx.close();
}

await writeFile(path.join(opts.out, "target-tier-raw.json"), JSON.stringify({
  what: "O5F §十B raw probe -- the live Target runtime under the exact "
      + "capture contexts, its compiled programs, its tier-predicate inputs "
      + "and the bundles it loaded.",
  target: opts.target, asset: opts.asset,
  contextRule: "mobile emulation exactly when w or h < 500 -- "
      + "o5r-measure.mjs's own rule",
  records,
}, null, 1));
console.log(`-> ${opts.out}/target-tier-raw.json`);
await browser.close();

File: scripts/v5/o5f-target-tier.py
#!/usr/bin/env python3
"""O5F §十B -- score the Target mobile sample tier.

Three facts, none inferred from viewport dimensions:

  1. THE PREDICATE, from the bundle. The archived card bundle carries the
     device-tier expression at a real byte offset; it is located here by
     regex over raw bytes and quoted. The tier is (pointer: coarse) OR
     hardwareConcurrency <= 6 OR deviceMemory <= 4 -- a device predicate,
     with no viewport term at all.
  2. THE LIVE RUNTIME. o5f-target-tier.mjs loaded the live page in the exact
     capture contexts and hooked shader-module creation; the card body
     program's refract-call count IS its sample count (the sealed compiled
     audit established that equivalence: \\brefract\\s*\\( appears exactly
     `samples` times). The predicate inputs were read live in each context.
  3. THE BUNDLE IDENTITY. Every live response containing
     "maxDispersionSamples" was hashed; it must equal the archived bundle,
     or fact 1 describes a different program than fact 2 ran.

Output: qa-v5/optics-o5f/target-mobile-tier.json
"""
from __future__ import annotations

import hashlib
import json
import re
import sys
from pathlib import Path

REPO = Path(__file__).resolve().parent.parent.parent
RAW = REPO / "artifacts/optics-o5f/target-tier/target-tier-raw.json"
BUNDLE = REPO / "artifacts/f27/bundles/03lo820gl57km.js"
OUT = REPO / "qa-v5/optics-o5f/target-mobile-tier.json"

OUR_SAMPLES_AT_CAPTURE = {
    # V5_BODY_SAMPLES[quality] at the captures' quality (adaptive off, high).
    "1440x900": 5, "390x844": 5, "844x390": 5, "700x700": 5,
}


def main() -> int:
    raw = json.loads(RAW.read_text())
    data = BUNDLE.read_bytes()
    bundle_sha = hashlib.sha256(data).hexdigest()

    # --- 1: the predicate, from raw bundle bytes ---------------------------
    pred = re.search(
        rb'matchMedia\("\(pointer: coarse\)"\)\.matches[^;]{0,400}?'
        rb'maxDispersionSamples:"low"===\w+\?3:1/0', data)
    if not pred:
        print("FATAL: predicate site not found in archived bundle")
        return 2
    excerpt = data[pred.start():pred.end()].decode("utf-8", errors="replace")
    tier_terms = {
        "pointerCoarse": 'window.matchMedia("(pointer: coarse)").matches',
        "cores": "navigator.hardwareConcurrency ?? 8, low when <= 6",
        "memory": "navigator.deviceMemory ?? 8, low when <= 4",
    }

    def predicted_tier(env):
        if env is None:
            return None
        coarse = bool(env.get("pointerCoarse"))
        cores = env.get("hardwareConcurrency") or 8
        mem = env.get("deviceMemory") or 8
        return "low" if (coarse or cores <= 6 or mem <= 4) else "high"

    # --- 2 + 3: per-context live readings ----------------------------------
    rows = []
    live_matches_archive = True
    for rec in raw["records"]:
        vp = rec["vp"]
        env = (rec.get("probe") or {}).get("envAtLoad")
        tier = predicted_tier(env)
        expected = {"low": 3, "high": 5}.get(tier)
        observed = None
        refract_calls = None
        prog = rec.get("bodyProgram")
        if prog:
            text = (RAW.parent / prog["file"]).read_text()
            refract_calls = len(re.findall(r"\brefract\s*\(", text))
            observed = refract_calls
        for b in rec.get("bundlesWithPredicate", []):
            if b["sha256"] != bundle_sha:
                live_matches_archive = False
        ours = OUR_SAMPLES_AT_CAPTURE.get(vp)
        rows.append({
            "vp": vp,
            "mobileEmulation": rec.get("mobileEmulation"),
            "loaded": rec.get("loaded"),
            "predicateInputs": env,
            "predictedTier": tier,
            "predictedSamples": expected,
            "observedRefractCalls": refract_calls,
            "observedSamples": observed,
            "agreement": (observed == expected
                          if observed is not None and expected is not None
                          else None),
            "ourCandidateSamples": ours,
            "mismatchWithOurs": (observed is not None and ours is not None
                                 and observed != ours),
            "bundles": rec.get("bundlesWithPredicate"),
            "bodyProgram": prog,
            "errorCount": rec.get("errorCount"),
        })

    p0 = next((r for r in rows if r["vp"] == "390x844"), None)
    landscape = next((r for r in rows if r["vp"] == "844x390"), None)

    # The counting-convention CONTROL. "refract-call count IS the sample
    # count" was established on OUR three build's compiled programs; before a
    # mobile-context reading may mean anything, the same equivalence has to be
    # shown to hold on the Target's compiled output. The desktop 1440x900
    # context is the known-tier case: its predicate inputs predict "high",
    # so its program must count exactly 5. If it does not, the convention
    # does not transfer and EVERY reading here is INSTRUMENT_UNREADABLE --
    # never a divergence, never a match.
    control = next((r for r in rows if r["vp"] == "1440x900"), None)
    convention_ok = (control is not None
                     and control["predictedTier"] == "high"
                     and control["observedRefractCalls"] == 5)
    counting_control = {
        "vp": "1440x900",
        "why": "desktop context with a known predicted tier (high -> 5); "
               "observing exactly 5 refract calls in the Target's live "
               "program validates the count-equals-samples convention on "
               "the Target's own compiled output, not just ours.",
        "predictedTier": None if control is None else control["predictedTier"],
        "observedRefractCalls": None if control is None
        else control["observedRefractCalls"],
        "validated": convention_ok,
    }
    readable = convention_ok and all(
        r["loaded"] and r["observedSamples"] is not None for r in rows)

    doc = {
        "what": "§十B -- the Target's mobile sample tier, proven from the "
                "live runtime and the bundle, never inferred from viewport "
                "dimensions.",
        "bundlePredicate": {
            "file": "artifacts/f27/bundles/03lo820gl57km.js",
            "byteOffset": pred.start(),
            "excerpt": excerpt,
            "terms": tier_terms,
            "reading": "low = (pointer: coarse) OR cores <= 6 OR "
                       "deviceMemory <= 4; maxDispersionSamples = low ? 3 : "
                       "Infinity; samples = min(5, maxDispersionSamples). "
                       "No viewport term exists in the predicate.",
            "archivedBundleSha256": bundle_sha,
        },
        "liveBundleMatchesArchive": live_matches_archive,
        "captureContextRule": raw["contextRule"],
        "countingConventionControl": counting_control,
        "rows": rows,
        "conclusion": {
            "p0": None if p0 is None else {
                "vp": "390x844",
                "targetSamplesObserved": p0["observedSamples"],
                "ourCandidateSamples": p0["ourCandidateSamples"],
                "sampleTierMismatch": p0["mismatchWithOurs"],
            },
            "landscape": None if landscape is None else {
                "vp": "844x390",
                "targetSamplesObserved": landscape["observedSamples"],
                "ourCandidateSamples": landscape["ourCandidateSamples"],
                "sampleTierMismatch": landscape["mismatchWithOurs"],
            },
        },
        "status": "READABLE" if readable else "INSTRUMENT_UNREADABLE",
        "unreadableNote": None if readable else
            ("the desktop control did not validate the refract-count "
             "convention on the Target's own program; no reading here can "
             "prove a tier." if not convention_ok else
             "a context that did not load or yielded no refract-bearing "
             "program cannot prove a tier; it is not converted into either "
             "verdict."),
    }
    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text(json.dumps(doc, indent=1))
    for r in rows:
        print(f"{r['vp']:>9}: coarse={((r['predicateInputs'] or {}).get('pointerCoarse'))} "
              f"predicted={r['predictedSamples']} observed={r['observedSamples']} "
              f"ours={r['ourCandidateSamples']} mismatch={r['mismatchWithOurs']}")
    print(f"counting convention (1440x900 control): "
          f"{'validated' if convention_ok else 'NOT validated'}")
    print(f"live bundle == archive: {live_matches_archive}")
    print(f"-> {OUT}  [{doc['status']}]")
    return 0


if __name__ == "__main__":
    sys.exit(main())
